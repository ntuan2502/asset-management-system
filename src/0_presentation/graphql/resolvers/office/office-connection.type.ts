import { ObjectType } from '@nestjs/graphql';
import { Paginated } from 'src/shared/dtos/pagination.dto';
import { OfficeType } from './office.type';

@ObjectType()
export class OfficeConnection extends Paginated(OfficeType) {}
