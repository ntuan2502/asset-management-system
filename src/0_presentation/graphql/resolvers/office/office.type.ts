import { Field, GraphQLISODateTime, ID, ObjectType } from '@nestjs/graphql';

@ObjectType('Office')
export class OfficeType {
  @Field(() => ID)
  id: string;

  @Field()
  name: string;

  @Field()
  internationalName: string;

  @Field()
  shortName: string;

  @Field()
  taxCode: string;

  @Field()
  address: string;

  @Field({ nullable: true })
  description: string;

  @Field(() => GraphQLISODateTime)
  createdAt: Date;

  @Field(() => GraphQLISODateTime)
  updatedAt: Date;
}
