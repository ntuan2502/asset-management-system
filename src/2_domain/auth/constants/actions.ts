export const ACTIONS = {
  CREATE: 'create',
  READ: 'read',
  UPDATE: 'update',
  DELETE: 'delete',
  MANAGE: 'manage', // Một quyền đặc biệt có nghĩa là "tất cả"
};
