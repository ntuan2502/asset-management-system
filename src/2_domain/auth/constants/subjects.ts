export const ENTITY_SUBJECTS = {
  USER: 'User',
  PERMISSION: 'Permission',
  ROLE: 'Role',
  ASSET: 'Asset',
  OFFICE: 'Office',
  DEPARTMENT: 'Department',
  STATUS_LABEL: 'StatusLabel',
  CATEGORY: 'Category',
  MANUFACTURER: 'Manufacturer',
  ATTRIBUTE: 'Attribute',
};
