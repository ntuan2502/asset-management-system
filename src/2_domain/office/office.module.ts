import { Module } from '@nestjs/common';
import { CqrsModule } from '@nestjs/cqrs';
import { OfficeResolver } from 'src/0_presentation/graphql/resolvers/office/office.resolver';
import { CreateOfficeHandler } from 'src/1_application/office/commands/handlers/create-office.handler';
import { OFFICE_REPOSITORY } from './repositories/office.repository.interface';
import { PrismaOfficeRepository } from 'src/3_infrastructure/persistence/prisma/repositories/office/prisma-office.repository';
import { OfficeProjector } from 'src/3_infrastructure/projection/office.projector';
import { SharedInfrastructureModule } from 'src/3_infrastructure/shared/shared-infrastructure.module';
import { GetAllOfficesHandler } from 'src/1_application/office/queries/handlers/get-all-offices.handler';
import { GetOfficeByIdHandler } from 'src/1_application/office/queries/handlers/get-office-by-id.handler';
import { OfficeAggregateRepository } from './repositories/office-aggregate.repository';
import { UpdateOfficeHandler } from 'src/1_application/office/commands/handlers/update-office.handler';
import { DeleteOfficeHandler } from 'src/1_application/office/commands/handlers/delete-office.handler';
import { RestoreOfficeHandler } from 'src/1_application/office/commands/handlers/restore-office.handler';

const CommandHandlers = [
  CreateOfficeHandler,
  UpdateOfficeHandler,
  DeleteOfficeHandler,
  RestoreOfficeHandler,
];
const QueryHandlers = [GetAllOfficesHandler, GetOfficeByIdHandler];
const Repositories = [
  { provide: OFFICE_REPOSITORY, useClass: PrismaOfficeRepository },
  OfficeAggregateRepository,
];
const Projectors = [OfficeProjector];

@Module({
  imports: [CqrsModule, SharedInfrastructureModule],
  providers: [
    OfficeResolver,
    ...CommandHandlers,
    ...QueryHandlers,
    ...Repositories,
    ...Projectors,
  ],
  exports: [OFFICE_REPOSITORY],
})
export class OfficeModule {}
