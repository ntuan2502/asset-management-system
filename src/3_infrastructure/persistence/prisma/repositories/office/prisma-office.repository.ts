import { Injectable } from '@nestjs/common';
import { PrismaService } from 'src/3_infrastructure/persistence/prisma/prisma.service';
import { IOfficeRepository } from 'src/2_domain/office/repositories/office.repository.interface';
import { OfficeMapper } from 'src/1_application/office/mappers/office.mapper';
import { PaginatedOffices } from 'src/2_domain/office/repositories/office.repository.interface';
import { OfficeAggregate } from 'src/2_domain/office/aggregates/office.aggregate';

@Injectable()
export class PrismaOfficeRepository implements IOfficeRepository {
  constructor(private readonly prisma: PrismaService) {}

  async isDuplicate(data: {
    internationalName: string;
    shortName: string;
    taxCode: string;
  }): Promise<boolean> {
    const count = await this.prisma.office.count({
      where: {
        OR: [
          { internationalName: data.internationalName },
          { shortName: data.shortName },
          { taxCode: data.taxCode },
        ],
        deletedAt: null,
      },
    });
    return count > 0;
  }

  async findById(id: string): Promise<OfficeAggregate | null> {
    const office = await this.prisma.office.findFirst({
      where: { id, deletedAt: null },
    });
    return office ? OfficeMapper.toDomain(office) : null;
  }

  async findAll(args: {
    page: number;
    limit: number;
  }): Promise<PaginatedOffices> {
    const { page, limit } = args;
    const skip = (page - 1) * limit;

    const [offices, totalCount] = await this.prisma.$transaction([
      this.prisma.office.findMany({
        where: { deletedAt: null },
        skip,
        take: limit,
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.office.count({ where: { deletedAt: null } }),
    ]);

    return {
      nodes: offices.map((office) => OfficeMapper.toDomain(office)),
      meta: { totalCount, page, limit },
    };
  }
}
