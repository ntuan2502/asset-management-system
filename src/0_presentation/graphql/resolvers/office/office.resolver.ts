import { Args, ID, Mutation, Query, Resolver } from '@nestjs/graphql';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { OfficeType } from './office.type';
import { CreateOfficeInput } from 'src/1_application/office/dtos/create-office.input';
import { CreateOfficeCommand } from 'src/1_application/office/commands/impl/create-office.command';
import { OfficeAggregate } from 'src/2_domain/office/aggregates/office.aggregate';
import { OfficeConnection } from './office-connection.type';
import { PaginationArgs } from 'src/shared/dtos/pagination-args.dto';
import { PaginatedOfficesResult } from 'src/1_application/office/queries/handlers/get-all-offices.handler';
import { GetAllOfficesQuery } from 'src/1_application/office/queries/impl/get-all-offices.query';
import { GetOfficeByIdQuery } from 'src/1_application/office/queries/impl/get-office-by-id.query';
import { UpdateOfficeInput } from 'src/1_application/office/dtos/update-office.input';
import { UpdateOfficeCommand } from 'src/1_application/office/commands/impl/update-office.command';
import { DeleteOfficeCommand } from 'src/1_application/office/commands/impl/delete-office.command';
import { UseGuards } from '@nestjs/common';
import { GqlAuthGuard } from 'src/2_domain/auth/guards/gql-auth.guard';
import { PermissionsGuard } from 'src/2_domain/auth/guards/permissions.guard';
import { CheckPermissions } from 'src/2_domain/auth/decorators/check-permissions.decorator';
import { ACTIONS } from 'src/2_domain/auth/constants/actions';
import { ENTITY_SUBJECTS } from 'src/2_domain/auth/constants/subjects';

@Resolver(() => OfficeType)
@UseGuards(GqlAuthGuard, PermissionsGuard)
export class OfficeResolver {
  constructor(
    private readonly commandBus: CommandBus,
    private readonly queryBus: QueryBus,
  ) {}

  @Mutation(() => OfficeType)
  @CheckPermissions({ action: ACTIONS.CREATE, subject: ENTITY_SUBJECTS.OFFICE })
  async createOffice(
    @Args('input') input: CreateOfficeInput,
  ): Promise<OfficeAggregate> {
    return this.commandBus.execute(new CreateOfficeCommand(input));
  }

  @Query(() => OfficeConnection, { name: 'offices' })
  @CheckPermissions({ action: ACTIONS.READ, subject: ENTITY_SUBJECTS.OFFICE })
  async getAllOffices(
    @Args() args: PaginationArgs,
  ): Promise<PaginatedOfficesResult> {
    return this.queryBus.execute(new GetAllOfficesQuery(args));
  }

  @Query(() => OfficeType, { name: 'office', nullable: true })
  @CheckPermissions({ action: ACTIONS.READ, subject: ENTITY_SUBJECTS.OFFICE })
  async getOfficeById(
    @Args('id', { type: () => ID }) id: string,
  ): Promise<OfficeAggregate | null> {
    return this.queryBus.execute(new GetOfficeByIdQuery(id));
  }

  @Mutation(() => OfficeType)
  @CheckPermissions({ action: ACTIONS.UPDATE, subject: ENTITY_SUBJECTS.OFFICE })
  async updateOffice(
    @Args('id', { type: () => ID }) id: string,
    @Args('input') input: UpdateOfficeInput,
  ): Promise<OfficeAggregate> {
    return this.commandBus.execute(new UpdateOfficeCommand(id, input));
  }

  @Mutation(() => Boolean)
  @CheckPermissions({ action: ACTIONS.DELETE, subject: ENTITY_SUBJECTS.OFFICE })
  async deleteOffice(
    @Args('id', { type: () => ID }) id: string,
  ): Promise<boolean> {
    await this.commandBus.execute(new DeleteOfficeCommand(id));
    return true;
  }
}
