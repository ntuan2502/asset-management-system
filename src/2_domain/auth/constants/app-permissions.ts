import { ACTIONS } from './actions';
import { SUBJECTS } from './subjects';

export type AppPermission = {
  action: string;
  subject: string;
};

export const ALL_APP_PERMISSIONS: AppPermission[] = [
  // Thêm quyền cho Permission (chỉ đọc)
  { action: ACTIONS.READ, subject: SUBJECTS.PERMISSION },

  // Quyền trên User
  { action: ACTIONS.CREATE, subject: SUBJECTS.USER },
  { action: ACTIONS.READ, subject: SUBJECTS.USER },
  { action: ACTIONS.UPDATE, subject: SUBJECTS.USER },
  { action: ACTIONS.DELETE, subject: SUBJECTS.USER },

  // Quyền trên Role
  { action: ACTIONS.CREATE, subject: SUBJECTS.ROLE },
  { action: ACTIONS.READ, subject: SUBJECTS.ROLE },
  { action: ACTIONS.UPDATE, subject: SUBJECTS.ROLE },
  { action: ACTIONS.DELETE, subject: SUBJECTS.ROLE },

  // Quyền trên Office
  { action: ACTIONS.CREATE, subject: SUBJECTS.OFFICE },
  { action: ACTIONS.READ, subject: SUBJECTS.OFFICE },
  { action: ACTIONS.UPDATE, subject: SUBJECTS.OFFICE },
  { action: ACTIONS.DELETE, subject: SUBJECTS.OFFICE },

  // Quyền trên Department
  { action: ACTIONS.CREATE, subject: SUBJECTS.DEPARTMENT },
  { action: ACTIONS.READ, subject: SUBJECTS.DEPARTMENT },
  { action: ACTIONS.UPDATE, subject: SUBJECTS.DEPARTMENT },
  { action: ACTIONS.DELETE, subject: SUBJECTS.DEPARTMENT },

  // Quyền trên StatusLabel
  { action: ACTIONS.CREATE, subject: SUBJECTS.STATUS_LABEL },
  { action: ACTIONS.READ, subject: SUBJECTS.STATUS_LABEL },
  { action: ACTIONS.UPDATE, subject: SUBJECTS.STATUS_LABEL },
  { action: ACTIONS.DELETE, subject: SUBJECTS.STATUS_LABEL },

  // Quyền trên Category
  { action: ACTIONS.CREATE, subject: SUBJECTS.CATEGORY },
  { action: ACTIONS.READ, subject: SUBJECTS.CATEGORY },
  { action: ACTIONS.UPDATE, subject: SUBJECTS.CATEGORY },
  { action: ACTIONS.DELETE, subject: SUBJECTS.CATEGORY },
];
