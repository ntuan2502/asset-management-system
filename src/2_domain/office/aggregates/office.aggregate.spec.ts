import { OfficeAggregate } from './office.aggregate';
import { OfficeCreatedEvent } from '../events/office-created.event';
import { OfficeUpdatedEvent } from '../events/office-updated.event';
import { DEPARTMENT_ERRORS } from 'src/shared/constants/error-messages.constants';

// Mock thư viện cuid2 để ID được tạo ra là cố định, giúp snapshot ổn định
jest.mock('@paralleldrive/cuid2', () => ({
  createId: jest.fn(() => 'mock-cuid-12345'),
}));

describe('OfficeAggregate', () => {
  // Dọn dẹp mock sau mỗi bài test
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('createOffice', () => {
    it('should apply an OfficeCreatedEvent and set initial state', () => {
      // --- Arrange ---
      const aggregate = new OfficeAggregate();
      const officeData = {
        name: 'Amata Vietnam JSC',
        internationalName: 'Amata Vietnam Joint Stock Company',
        shortName: 'AMATAVN',
        taxCode: '0123456789',
        address: '123 Main St, HCMC',
        description: 'Headquarters',
      };

      // Đóng băng thời gian để snapshot ổn định
      jest.useFakeTimers().setSystemTime(new Date('2025-01-01T10:00:00.000Z'));

      // --- Act ---
      aggregate.createOffice(officeData);

      // --- Assert ---
      const uncommittedEvents = aggregate.getUncommittedEvents();
      const createdEvent = uncommittedEvents[0];

      // 1. Khẳng định tường minh (Explicit Assertions)
      expect(uncommittedEvents).toHaveLength(1);
      expect(createdEvent).toBeInstanceOf(OfficeCreatedEvent);
      expect((createdEvent as OfficeCreatedEvent).id).toBe('mock-cuid-12345');
      expect((createdEvent as OfficeCreatedEvent).name).toBe(officeData.name);

      // 2. Khẳng định bằng Snapshot
      //    Chụp ảnh sự kiện để đảm bảo tất cả các trường khác đều đúng
      expect(createdEvent).toMatchSnapshot();

      //    Chụp ảnh trạng thái của aggregate để đảm bảo `onOfficeCreatedEvent` hoạt động đúng
      expect(aggregate).toMatchSnapshot();

      // Dọn dẹp mock thời gian
      jest.useRealTimers();
    });
  });

  describe('updateOffice', () => {
    // Tạo một hàm helper để tái sử dụng việc thiết lập trạng thái
    const createExistingOffice = (): OfficeAggregate => {
      const aggregate = new OfficeAggregate();
      // Gán trạng thái trực tiếp, mô phỏng một office đã tồn tại
      aggregate.id = 'office-123';
      aggregate.name = 'Old Name';
      aggregate.internationalName = 'Old International Name';
      aggregate.shortName = 'OLD';
      aggregate.taxCode = '12345';
      aggregate.address = 'Old Address';
      aggregate.description = 'Old Description';
      aggregate.deletedAt = null;
      aggregate.createdAt = new Date('2025-01-01T00:00:00.000Z');
      aggregate.updatedAt = new Date('2025-01-01T00:00:00.000Z');
      aggregate.version = 1;
      return aggregate;
    };

    it('should apply an OfficeUpdatedEvent when data changes', () => {
      // --- Arrange ---
      const aggregate = createExistingOffice();
      const updatePayload = {
        name: 'New Name',
        description: 'New Description',
      };

      // Đóng băng thời gian cho `updatedAt` trong sự kiện
      jest.useFakeTimers().setSystemTime(new Date('2025-02-02T11:00:00.000Z'));

      // --- Act ---
      aggregate.updateOffice(updatePayload);

      // --- Assert ---
      const uncommittedEvents = aggregate.getUncommittedEvents();
      expect(uncommittedEvents).toHaveLength(1);
      expect(uncommittedEvents[0]).toBeInstanceOf(OfficeUpdatedEvent);

      // Chụp ảnh sự kiện update (chỉ chứa các thay đổi)
      expect(uncommittedEvents[0]).toMatchSnapshot('event');
      // Chụp ảnh trạng thái cuối cùng của aggregate
      expect(aggregate).toMatchSnapshot('aggregate state');

      jest.useRealTimers();
    });

    it('should not apply an event if no data has changed', () => {
      // --- Arrange ---
      const aggregate = createExistingOffice();
      // Gửi lại dữ liệu giống hệt trạng thái hiện tại
      const updatePayload = { name: 'Old Name' };

      // --- Act ---
      aggregate.updateOffice(updatePayload);

      // --- Assert ---
      // Mong đợi không có sự kiện nào được phát ra
      expect(aggregate.getUncommittedEvents()).toHaveLength(0);
    });

    it('should throw an error when updating a deleted office', () => {
      // --- Arrange ---
      const aggregate = createExistingOffice();
      aggregate.deletedAt = new Date(); // Giả lập trạng thái đã bị xóa

      // --- Act & Assert ---
      const action = () => aggregate.updateOffice({ name: 'New Name' });
      expect(action).toThrow(DEPARTMENT_ERRORS.CANNOT_UPDATE_DELETED);
    });
  });
});
