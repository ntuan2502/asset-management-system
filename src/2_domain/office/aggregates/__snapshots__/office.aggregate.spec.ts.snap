// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfficeAggregate createOffice should apply an OfficeCreatedEvent and set initial state 1`] = `
OfficeCreatedEvent {
  "address": "123 Main St, HCMC",
  "createdAt": 2025-01-01T10:00:00.000Z,
  "description": "Headquarters",
  "id": "mock-cuid-12345",
  "internationalName": "Amata Vietnam Joint Stock Company",
  "name": "Amata Vietnam JSC",
  "shortName": "AMATAVN",
  "taxCode": "0123456789",
}
`;

exports[`OfficeAggregate createOffice should apply an OfficeCreatedEvent and set initial state 2`] = `
OfficeAggregate {
  "address": "123 Main St, HCMC",
  "aggregateType": "Office",
  "createdAt": 2025-01-01T10:00:00.000Z,
  "deletedAt": null,
  "description": "Headquarters",
  "id": "mock-cuid-12345",
  "internationalName": "Amata Vietnam Joint Stock Company",
  "name": "Amata Vietnam JSC",
  "shortName": "AMATAVN",
  "taxCode": "0123456789",
  "updatedAt": 2025-01-01T10:00:00.000Z,
  "version": 1,
  Symbol(): false,
  Symbol(): [
    OfficeCreatedEvent {
      "address": "123 Main St, HCMC",
      "createdAt": 2025-01-01T10:00:00.000Z,
      "description": "Headquarters",
      "id": "mock-cuid-12345",
      "internationalName": "Amata Vietnam Joint Stock Company",
      "name": "Amata Vietnam JSC",
      "shortName": "AMATAVN",
      "taxCode": "0123456789",
    },
  ],
}
`;

exports[`OfficeAggregate updateOffice should apply an OfficeUpdatedEvent when data changes: aggregate state 1`] = `
OfficeAggregate {
  "address": "Old Address",
  "aggregateType": "Office",
  "createdAt": 2025-01-01T00:00:00.000Z,
  "deletedAt": null,
  "description": "New Description",
  "id": "office-123",
  "internationalName": "Old International Name",
  "name": "New Name",
  "shortName": "OLD",
  "taxCode": "12345",
  "updatedAt": 2025-02-02T11:00:00.000Z,
  "version": 2,
  Symbol(): false,
  Symbol(): [
    OfficeUpdatedEvent {
      "address": undefined,
      "description": "New Description",
      "id": "office-123",
      "internationalName": undefined,
      "name": "New Name",
      "shortName": undefined,
      "taxCode": undefined,
      "updatedAt": 2025-02-02T11:00:00.000Z,
    },
  ],
}
`;

exports[`OfficeAggregate updateOffice should apply an OfficeUpdatedEvent when data changes: event 1`] = `
OfficeUpdatedEvent {
  "address": undefined,
  "description": "New Description",
  "id": "office-123",
  "internationalName": undefined,
  "name": "New Name",
  "shortName": undefined,
  "taxCode": undefined,
  "updatedAt": 2025-02-02T11:00:00.000Z,
}
`;
